### PR TITLE
Google Stackdriver Utilization Policy

### PR DESCRIPTION
**Pull request by rshade**
### Description
Google Stackdriver Policy

### Issues Resolved

[List any existing issues this PR resolves]

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

File: cost/google/instance_stackdriver_utilization/google_instance_stackdriver_utilization.pt [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+# Datasources
+###############################################################################
+
+datasource "ds_compute_utilization" do
+  request do
+    auth $auth_google
+    host "monitoring.googleapis.com"
+    path join(["/v3/projects/",to_s($param_google_project),"/timeSeries/"])
+    query "filter", "metric.type=\"compute.googleapis.com/instance/cpu/utilization\""
+    query "interval.endTime", "2019-05-03T00:00:00.000000Z"
```
Comment (srpomeroy) [inline cost/google/instance_stackdriver_utilization/google_instance_stackdriver_utilization.pt]: Are startTime and endTime supposed to be hardcoded?

File: cost/google/instance_stackdriver_utilization/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+## Google Instance StackDriver Utilization Policy
+
+### What it does
+
+This Policy Template gets StackDriver data for instances in google cloud
+
+### Cloud Management Required Permissions/AWS Required Permissions
```
Comment (srpomeroy) [inline cost/google/instance_stackdriver_utilization/README.md]: Not AWS

File: cost/google/instance_stackdriver_utilization/README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
+## Google Instance StackDriver Utilization Policy
+
+### What it does
+
+This Policy Template gets StackDriver data for instances in google cloud
```
Comment (srpomeroy) [inline cost/google/instance_stackdriver_utilization/README.md]: Should be Google not google

File: cost/google/instance_stackdriver_utilization/README.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+### Cloud Management Required Permissions/AWS Required Permissions
+- Cloud Management - The `credential_viewer`,`observer` roles
+- Cloud Management - The `policy_designer`, `policy_manager` & `policy_publisher` roles
+- Google - The `Monitoring Viewer ` Role
+
+### Functional Details
+
+- This policy uses RightScale to get a list of instances, it then polls google StackDriver for metrics for instance performance and delivers a report. If you get an **N/A** in a field you will need to install the [StackDriver Agent](https://cloud.google.com/monitoring/agent/install-agent) on the instance to get those metrics.
```
Comment (srpomeroy) [inline cost/google/instance_stackdriver_utilization/README.md]: Google not google

File: cost/google/instance_stackdriver_utilization/google_instance_stackdriver_utilization.pt [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+  request do
+    run_script $js_request_compute_utilization, $param_google_project
+  end
+  result do
+    encoding "json"
+    collect jmes_path(response, "timeSeries[*]") do
+      field "label_instance_name", jmes_path(col_item, "metric.labels.instance_name")
+      field "instance_id", jmes_path(col_item, "resource.labels.instance_id")
+      field "zone", jmes_path(col_item, "resource.labels.zone")
+      # field "avg_points", jmes_path(col_item,"points[].[avg(@),value.doubleValue]")
```
Comment (srpomeroy) [inline cost/google/instance_stackdriver_utilization/google_instance_stackdriver_utilization.pt]: Is this needed?

File: cost/google/instance_stackdriver_utilization/google_instance_stackdriver_utilization.pt [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+  end
+end
+
+datasource "ds_disk_utilization" do
+  request do
+    auth $auth_google
+    host "monitoring.googleapis.com"
+    path join(["/v3/projects/",to_s($param_google_project),"/timeSeries/"])
+    query "filter", "metric.type=\"agent.googleapis.com/disk/percent_used\" AND metric.label.state=used"
+    query "interval.endTime", "2019-05-03T00:00:00.000000Z"
```
Comment (srpomeroy) [inline cost/google/instance_stackdriver_utilization/google_instance_stackdriver_utilization.pt]: Should start and end be hardcoded?

File: cost/google/instance_stackdriver_utilization/google_instance_stackdriver_utilization.pt [hunk 7]
```diff
@@ -0,0 +1,281 @@
+name "Google Instance StackDriver Utilization"
+rs_pt_ver 20180301
+type "policy"
+short_description "Checks Google Regions and reports on StackDriver Utilization. See the [README](https://github.com/rightscale/policy_templates/tree/master/cost/google/instance_stackdriver_utilization/) and [docs.rightscale.com/policies](http://docs.rightscale.com/policies/) to learn more."
```
Comment (srpomeroy) [inline cost/google/instance_stackdriver_utilization/google_instance_stackdriver_utilization.pt]: Should this be "Google Instances"? We should also include a note about the timeframe calculated.
Example: "Reports on Google instances StackDriver Utilization over the past 30 days..."

File: cost/google/instance_stackdriver_utilization/README.md (hunk 3 above)
Comment (srpomeroy) [reply cost/google/instance_stackdriver_utilization/README.md]: Should also include details around the timeframe used to collect the metrics.

File: cost/google/instances_stackdriver_utilization/google_instances_stackdriver_utilization.pt [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+    query "pageToken"
+  end
+end
+
+###############################################################################
+# Datasources
+###############################################################################
+
+datasource "ds_compute_utilization" do
+  request do
```
Comment (avinashbhashyam-rs) [inline cost/google/instances_stackdriver_utilization/google_instances_stackdriver_utilization.pt]: From what I understand you need start_time and end_time query parameters to make request and that's the reason for using java script to make request. The easier solution would be to use javascript to just calculate start and end time and refer to them as query parameters in http datasource. That allows you to paginate the results.

File: cost/google/instances_stackdriver_utilization/google_instances_stackdriver_utilization.pt [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+  parameters "ds_compute_utilization", "ds_memory_utilization", "ds_disk_utilization"
+  code <<-EOS
+     results = []
+     for ( i =0; i < ds_compute_utilization.length; i++ ){
+       var instance_id = ds_compute_utilization[i].instance_id
+       var points = ds_compute_utilization[i].points
+       var memory_record = _.find(ds_memory_utilization, function(record) { return record.instance_id == instance_id; })
+       if (memory_record != null && memory_record !== undefined) {
+        var memory_points = memory_record.points
+        var memory_average = parseFloat(memory_points.reduce(function(acc, val) { return acc + val; })/memory_points.length).toFixed(2);
```
Comment (avinashbhashyam-rs) [inline cost/google/instances_stackdriver_utilization/google_instances_stackdriver_utilization.pt]: you are calculation avg_memory points while collecting the data_source `field "avg_disk_points", jmes_path(col_item, "points[*].value.doubleValue | avg(@)")`. Why are you not just referring to it?

File: cost/google/instances_stackdriver_utilization/google_instances_stackdriver_utilization.pt (hunk 8 above)
Comment (rshade) [reply cost/google/instances_stackdriver_utilization/google_instances_stackdriver_utilization.pt]: so just a datasource that returns a time object?

File: cost/google/instances_stackdriver_utilization/google_instances_stackdriver_utilization.pt (hunk 9 above)
Comment (rshade) [reply cost/google/instances_stackdriver_utilization/google_instances_stackdriver_utilization.pt]: I added those, but did not know what the output of the jmes_path calculation would be if points did not exist. I will update today.

**Pull request merged**